arclines/holy/patterns.py

```diff
@@ -222,3 +223,182 @@ def score_quad_matches(fidx):
             scores.append('Amb')
     # Return
     return scores
+
+
+def triangles(detlines, linelist, detsrch, lstsrch, npixels, pixtol):
+    """
+    Parameters
+    ----------
+    detlines : ndarray
+      list of detected lines in pixels (sorted, increasing)
+    linelist : ndarray
+      list of lines that should be detected (sorted, increasing)
+    detsrch : int
+      Number of consecutive elements in detlines to use to create a pattern (-1 means all lines in detlines)
+    lstsrch : int
+      Number of consecutive elements in linelist to use to create a pattern (-1 means all lines in detlines)
+    pixtol : float
+      tolerance that is used to determine if a match is successful (in units of pixels)
+
+    Returns
+    -------
+    dindex : ndarray
+      Index array of all detlines used in each triangle
+    lindex : ndarray
+      Index array of the assigned line to each index in dindex
+    wvcen : ndarray
+      central wavelength of each triangle
+    disps : ndarray
+      Dispersion of each triangle (angstroms/pixel)
+
+    """
+
+    nptn = 3  # Number of lines used to create a pattern
+
+    sz_d = detlines.size
+    sz_l = linelist.size
+
+    # Count the number of detlines patterns that will be created
+    cntdet = 0
+    dup = 0
+    for d in range(detsrch-nptn+1):
+        dup += d+1
+        if d == detsrch-nptn:
+            cntdet += dup*(sz_d-detsrch+1)
+        else:
+            cntdet += dup
+
+    # Count the number of linelist patterns that will be created
+    cntlst = 0
+    lup = 0
+    for l in range(lstsrch-nptn+1):
+        lup += l+1
+        if l == lstsrch-nptn:
+            cntlst += lup*(sz_l-lstsrch+1)
+        else:
+            cntlst += lup
+
+    lindex = np.zeros((cntdet*cntlst, nptn), dtype=np.int)
+    dindex = np.zeros((cntdet*cntlst, nptn), dtype=np.int)
+    wvcen = np.zeros((cntdet*cntlst), dtype=np.float)
+    disps = np.zeros((cntdet*cntlst), dtype=np.float)
+
+    # Test each detlines combination
+    cntdet = 0
+    for d in range(0, sz_d-nptn+1):
+        dup = d + detsrch
+        if dup > sz_d:
+            dup = sz_d
+        if detsrch == -1:
+            dup = sz_d
+        for dd in range(d+nptn-1, dup):
```

Comment:
Loving it.

Am going to think about a way to avoid one of these FOR loops but
we might be stuck with them..

Reply:
Yes, I agree - this is the slow part of the method, but in cython it is 100x faster.

Reply:
I should add, in the current version of the code, this is only used when the cython code cannot be compiled. So, cython is not technically a requirement, but it does offer a speedup.